### ISSUE TITLE
Don't support unpack of vector<bool>

### ISSUE THREAD

**Issue by hefengrren** [msgpack/msgpack-c#198]
vector<bool> could be packed OK, but could not be unpacked (unable to compile).

Possible solution: https://gist.github.com/hefengrren/081549f26e3b3b9eae1d.

**Comment by redboltz** [comment]
@hefengrren , hefenThank you for reporting the issue. I will implement std::vector<bool> support based on your approach.

**Comment by redboltz** [comment]
I chose an overloading approach instead of specialization.
http://www.gotw.ca/publications/mill17.htm

**Comment by hefengrren** [comment]
Good to know non-template function matches first.

But, your implementation does not check `p`, which is checked in other places.
I'm not sure of the expectation/design. Which goes to the end first, `p` or `it`, or both at the same time?

**Comment by redboltz** [comment]
p is a pointer of the array. The array has the following format:
https://github.com/msgpack/msgpack/blob/master/spec.md#formats-array

The array must have o.via.array.size elements. If it doesn't, unpack never successfully finished.


Then, v.resize(o.via.array.size) is called. 

https://github.com/msgpack/msgpack-c/pull/199/files#diff-976e6509b291276893ca5949e7f68d51R31

The size of v is the same as o.via.array.size even if v has a bigger size than o.via.array.size.
See http://www.cplusplus.com/reference/vector/vector/resize/

If v cannot be resized, exception would be thrown.
Hence, the size of v and o.via.array.size is the same after resize has successfully finished. And the memory from o.via.array.ptr to o.via.array.ptr + o.via.array.size is filled with msgpack objects.
In other words, p and it reach end at the same time.

**Comment by hefengrren** [comment]
Thanks for the detailed explanation!
